parser-typechecker/src/Unison/Codebase/Branch.hs

```diff
@@ -211,80 +220,163 @@ branch0 terms types children edits =
+              typeMetadata =
+                map
+                  (\(r, n, t, v) -> (r, Name.fromReverseSegments (n NonEmpty.:| reversePrefix), t, v))
+                  (deepMetadataHelper (_types b0))
+          children <- deepChildrenHelper e
+          go (work <> children) (typeMetadata <> acc)
 
 -- | Derive the 'deepPaths' field of a branch.
-deriveDeepPaths :: Branch0 m -> Branch0 m
+deriveDeepPaths :: forall m. Branch0 m -> Branch0 m
 deriveDeepPaths branch =
-  branch {deepPaths = makeDeepPaths (nonEmptyChildren branch)}
+  branch {deepPaths = makeDeepPaths branch}
   where
-    makeDeepPaths :: Map NameSegment (Branch m) -> Set Path
-    makeDeepPaths children =
-      Set.mapMonotonic Path.singleton (Map.keysSet children) <> ifoldMap go children
+    makeDeepPaths :: Branch0 m -> Set Path
+    makeDeepPaths branch = State.evalState (go (Seq.singleton ([], 0, branch)) mempty) Set.empty
       where
-        go :: NameSegment -> Branch m -> Set Path
-        go n b =
-          Set.map (Path.cons n) (deepPaths $ head b)
+        go :: Seq (DeepChildAcc m) -> Set Path -> DeepState m (Set Path)
+        go Seq.Empty acc = pure acc
+        go (e@(reversePrefix, _, b0) Seq.:<| work) acc = do
+          let paths :: Set Path
+              paths =
+                if isEmpty0 b0
+                  then Set.empty
+                  else (Set.singleton . Path . Seq.fromList . reverse) reversePrefix
+          children <- deepChildrenHelper e
+          go (work <> children) (paths <> acc)
 
 -- | Derive the 'deepEdits' field of a branch.
-deriveDeepEdits :: Branch0 m -> Branch0 m
+deriveDeepEdits :: forall m. Branch0 m -> Branch0 m
 deriveDeepEdits branch =
-  branch {deepEdits = makeDeepEdits (_edits branch) (nonEmptyChildren branch)}
+  branch {deepEdits = makeDeepEdits branch}
   where
-    makeDeepEdits :: Map NameSegment (EditHash, m Patch) -> Map NameSegment (Branch m) -> Map Name EditHash
-    makeDeepEdits edits children =
-      Map.mapKeysMonotonic Name.fromSegment (Map.map fst edits) <> ifoldMap go children
+    makeDeepEdits :: Branch0 m -> Map Name EditHash
+    makeDeepEdits branch = State.evalState (go (Seq.singleton ([], 0, branch)) mempty) Set.empty
       where
-        go :: NameSegment -> Branch m -> Map Name EditHash
-        go n b =
-          Map.mapKeys (Name.cons n) (deepEdits $ head b)
+        go :: (Seq (DeepChildAcc m)) -> Map Name EditHash -> DeepState m (Map Name EditHash)
+        go Seq.Empty acc = pure acc
+        go (e@(reversePrefix, _, b0) Seq.:<| work) acc = do
+          let edits :: Map Name EditHash
+              edits =
+                Map.mapKeysMonotonic
+                  (Name.fromReverseSegments . (NonEmpty.:| reversePrefix))
+                  (fst <$> _edits b0)
+          children <- deepChildrenHelper e
+          go (work <> children) (edits <> acc)
+
+type DeepState m = State (Set (NamespaceHash m))
```

Comment:
Opinion: these aliases (maybe all aliases) hurt more than they help. But feel free to ignore.

Reply:
What is this set, though?

Reply:
It's the set of visited namespace hashes.